catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/IngestCommand.java

```diff
@@ -506,84 +453,97 @@ private void moveToFailedIngestDirectory(File source) {
+                        ignoreCount.incrementAndGet();
+                    } else {
+                        String extension = file.getName();
 
-        private ArrayList<Metacard> metacards = new ArrayList<>(batchSize);
+                        if (extension.contains(".")) {
+                            int x = extension.indexOf('.');
+                            extension = extension.substring(x);
+                        }
 
-        private List<String> ignoreList;
+                        if (ignoreList != null && (ignoreList.contains(extension)
+                                || ignoreList.contains(file.getName()))) {
+                            ignoreCount.incrementAndGet();
+                            printProgressAndFlush(start,
+                                    fileCount.get(),
+                                    ingestCount.get() + ignoreCount.get());
+                        } else {
+                            Metacard result;
+                            try {
+                                result = readMetacard(file);
+                            } catch (IngestException e) {
+                                result = null;
+                                logIngestException(e, file);
+                                if (failedIngestDirectory != null) {
+                                    moveToFailedIngestDirectory(file);
+                                }
+                                printErrorMessage(
+                                        "Failed to ingest file [" + file.getAbsolutePath() + "].");
+                                if (INGEST_LOGGER.isWarnEnabled()) {
+                                    INGEST_LOGGER.warn("Failed to ingest file [{}].",
+                                            file.getAbsolutePath());
+                                }
+                            }
 
-        private List<File> files;
+                            if (result != null) {
+                                metacardQueue.add(result);
+                            }
+                        }
+                    }
+                });
+        doneBuildingQueue.set(true);
+    }
 
-        public IngestRunnable(long start, CatalogFacade catalog, List<String> ignoreList,
-                List<File> files) {
-            this.start = start;
-            this.catalog = catalog;
-            this.ignoreList = ignoreList;
-            this.files = files;
-        }
+    private void submitToCatalog(ScheduledExecutorService batchScheduler,
+            ExecutorService executorService, Queue<Metacard> metacardQueue, CatalogFacade catalog,
+            int batchSize, long start) {
 
-        @Override
-        public void run() {
-            long failed = 0;
-            for (File file : files) {
-                String extension = file.getName();
+        batchScheduler.scheduleWithFixedDelay(new Runnable() {
+            @Override
+            public void run() {
+                int queueSize = metacardQueue.size();
```

Comment:
Be aware that this .size() call is going to traverse the entire queue every time it is called. You might see better performance by using an ArrayBlockingQueue instead. Then you can also just use the .drainTo() method to pull off all the items you want instead of looping through them.